architecture/CACHING.md

```diff
@@ -0,0 +1,90 @@
+## Caching Middleware Architecture
+
+Package `cachemdw` is responsible for caching EVM requests and provides corresponding middleware
+
+package can work with any underlying storage which implements simple `cache.Cache` interface
+
+package provides two different middlewares:
+- `IsCachedMiddleware` (should be run before proxy middleware)
+- `CachingMiddleware`  (should be run after proxy middleware)
+
+`IsCachedMiddleware` is responsible for setting response in the context if it's in the cache
+
+`CachingMiddleware` is responsible for caching response by taking a value from context (should be set by `ProxyMiddleware`) and setting in the cache
+
+## CachingMiddleware
+
+`CachingMiddleware` returns kava-proxy-service compatible middleware which works in the following way:
+- tries to get decoded request from context (previous middleware should set it)
+- checks few conditions:
+  - if request isn't already cached
+  - if request is cacheable
+  - if response is present in context
+- if all above is true - caches the response
+- calls next middleware
```

Comment:
this always need to happen, regardless of whether redis is down, or the caching middleware has bugs. The service needs to practice graceful degradation (as outlined here https://github.com/Kava-Labs/kava-proxy-service/blob/main/architecture/ARCHITECTURE.md?plain=1#L35 as job #1 for the proxy service)

Reply:
@evgeniy-scherbina 
- what types of testing / validation have you done (manually or with automated tests) that even if the cache middleware encounters an error it won't prevent the
- can you please add a config (parsed from an environment variable) that allows for disabling of the caching middleware so that [when](https://martinfowler.com/bliki/TwoHardThings.html) there is a bug in the caching middleware the on-call can disable it without having to fix the bug first (allowing requests to continue being served from the backend hosts as they are currently in production)

Reply:
@galxy25 
- `what types of testing / validation have you done (manually or with automated tests) that even if the cache middleware encounters an error it won't prevent the...` - good point, going to add corresponding e2e-tests
- if you want to disable cache, you just have to provide `CACHE_ENABLED=false`